xlators/mgmt/glusterd/src/glusterd-svc-helper.c

```diff
@@ -862,65 +861,50 @@ __glusterd_send_svc_configure_req(glusterd_svc_t *svc, int flags,
     req_size = xdr_sizeof((xdrproc_t)xdr_gd1_mgmt_brick_op_req, req);
     iobuf = iobuf_get2(rpc->ctx->iobuf_pool, req_size);
     if (!iobuf) {
-        goto *errlbl;
+        goto err;
     }
-    errlbl = &&maybe_free_iobuf;
 
     iov.iov_base = iobuf->ptr;
     iov.iov_len = iobuf_pagesize(iobuf);
 
     iobref = iobref_new();
     if (!iobref) {
-        goto *errlbl;
+        goto err;
     }
-    errlbl = &&free_iobref;
 
     iobref_add(iobref, iobuf);
-    /*
-     * Drop our reference to the iobuf.  The iobref should already have
-     * one after iobref_add, so when we unref that we'll free the iobuf as
-     * well.  This allows us to pass just the iobref as frame->local.
-     */
-    iobuf_unref(iobuf);
-    /* Set the pointer to null so we don't free it on a later error. */
-    iobuf = NULL;
 
     /* Create the xdr payload */
     ret = xdr_serialize_generic(iov, req, (xdrproc_t)xdr_gd1_mgmt_brick_op_req);
     if (ret == -1) {
-        goto *errlbl;
+        goto err;
     }
     iov.iov_len = ret;
 
     /* Send the msg */
     GF_ATOMIC_INC(conf->blockers);
     ret = rpc_clnt_submit(rpc, &gd_brick_prog, op, cbkfn, &iov, 1, NULL, 0,
                           iobref, frame, NULL, 0, NULL, 0, NULL);
-    if (dict)
-        dict_unref(dict);
-    GF_FREE(volfile_content);
-    if (spec_fd >= 0)
-        sys_close(spec_fd);
-    return ret;
 
-free_iobref:
-    iobref_unref(iobref);
-maybe_free_iobuf:
+    frame = NULL;
```

Comment:
If `ret` contains an error, we should destroy `frame`, and probably decrement `conf->blockers`.

Reply:
frame is being destroyed in line: 906 if ret contains an error, I have added this line so as to avoid the double free scenario, as in positive scenario frame is being destroyed in the rpc_clnt_submit().
And, I think STACK_DESTROY(), takes care of doing GF_ATOMIC_DEC()?

Reply:
frame is also destroyed in the function glusterd_svc_attach_cbk() defined in line:651 which is passed to the function rpc_clnt_submit() as cbkfn as argument. The func glusterd_svc_attach_cbk destroys frame but since it's not being passed as a reference it won't affect here. That's the reason I am setting frame = NULL here.

Reply:
@xhernandez any comments for this?
I checked even `conf->blockers` is being decremented in the function `glusterd_svc_attach_brick()`.

Reply:
I think we are not talking about the same. The problem is that if `rpc_clnt_submit()` fails (`ret < 0`), it may not call `cbkfn`, so who will destroy `frame` and decrement `conf->blockers` ?

Reply:
Ohh, so the thing is code in the `rpc_clnt_submit()` is such that it calls the `cbkfn` function in the later part of the code only, which is `if(frame && (ret == -1))`, you can check the same in line:1755 in file `rpc/rpc-lib/src/rpc-clnt.c` . That's the reason I said the frame will be freed when it returns from the `rpc_clnt_submit()` if anything goes wrong.

Reply:
This is only true if `rpcreq` has been created. Otherwise it won't call `cbkfn`.

Reply:
Yeah, got it.
So, will it be good if I change the value of ret to `ret = -2` for cases(here it is 2 such case) before the creation of `rpcreq` is a success and then add a check in the caller func like `if (ret != -2) then frame = NULL` and do it. Otherwise, if I am not mistaken the `cbkfn` will be called and it would free `frame`. Or is there anything else that can be done?

Reply:
Yes, you could do something like that. It's not the best option IMO, but touching these functions is delicate, so this approach seems less intrusive. After the change, make sure that all callers correctly handle the new error value.

Reply:
Cool, I will make the changes.
But, I was thinking that previously too it worked this way but we didn't have any issues, or was it overlooked?

Reply:
It was a bug. Fortunately memory allocation almost never fails.

Reply:
Ohh, okay.